Implement SelectionArea single click/tap gestures

Renzo-Olivares opened the pull request:
This change collapses the selection at the clicked/tapped location on single click down for desktop platforms, and on single click/tap up for mobile platforms to match native.

This is a change from how `SelectionArea` previously worked. Before this change a single click down would clear the selection. From observing a native browser it looks like when tapping on static text the selection is not cleared but collapsed. A user can still attain the selection from static text using the `window.getSelection` API.

https://jsfiddle.net/juepasn3/11/ You can try this demo out here to observe this behavior yourself. When clicking on static text the selection will change.

This change also allows `Paragraph.selections` to return selections that are collapsed. This for testing purposes to confirm where the selection has been collapsed.

Partially fixes: #129583

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

Renzo-Olivares committed:
single click down should collapse selection on desktop/single click down/up should collapse the selection on mobile

Renzo-Olivares committed:
Paragraph.selection should not return empty for collapsed selections

Renzo-Olivares committed:
Fix test

Renzo-Olivares committed:
fix test

Renzo-Olivares committed:
remove logs

Renzo-Olivares committed:
fix tests

Renzo-Olivares committed:
should hide toolbar on touch tap up

chunhtai commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -978,7 +978,7 @@ void main() {
         ),
       );
-      expect(paragraph.selections.length, 0); // how []are you
+      expect(paragraph.selections.length, 1); // how []are you
```
Can you add a check that the selection is collapsed in the expected location?

chunhtai commented on packages/flutter/test/widgets/selectable_region_test.dart:
```diff
@@ -532,13 +537,13 @@ void main() {
       await gesture.up();
       await gesture.down(textOffsetToPosition(paragraph, 5));
       await tester.pumpAndSettle();
-      expect(paragraph.selections.isEmpty, isTrue);
```
Maybe check the selection is collapsed at the right place?

chunhtai reviewed (approved):
LGTM, thanks for fixing this!

Renzo-Olivares committed:
address comments

justinmc commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -485,7 +489,17 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
+          case TargetPlatform.android:
+          case TargetPlatform.fuchsia:
+          case TargetPlatform.iOS:
+            // On mobile platforms the selection is set on tap up.
```
Is this really a platform thing or is it a mouse vs. tap thing?

Renzo-Olivares replied:
Yeah this is a platform thing it seems like. I tried my iPad with the trackpad and keyboard and it doesn't set the selection until tap up. I also tried this on my pixel 7 with a mouse with the same results.

justinmc replied:
Alright, thank you for checking!

justinmc commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -949,6 +979,22 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
+  ///  * [_clearSelection], which clear the ongoing selection.
+  ///  * [_selectWordAt], which selects a whole word at the location.
+  ///  * [selectAll], which selects the entire content.
+  void _selectPositionAt({required Offset offset}) {
```
The name sounds to me like it will not result in a collapsed selection.  Maybe collapseSelectionAt or something like that?

justinmc commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -512,6 +526,24 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
+  void _handleMouseTapUp(TapDragUpDetails details) {
+    switch (_getEffectiveConsecutiveTapCount(details.consecutiveTapCount)) {
+      case 1:
+        switch (defaultTargetPlatform) {
```
Same question here.

Renzo-Olivares replied:
https://github.com/flutter/flutter/pull/132682#discussion_r1298865219

justinmc reviewed (commented):
I left two comments where I'm wondering if the behavior is based on platform, or mouse vs. tap.  Can you try plugging a physical mouse into a phone to confirm?  If you don't already know the answer.

Otherwise looks good!

Renzo-Olivares committed:
Address comments

Renzo-Olivares committed:
Address comments

Renzo-Olivares commented:
Thanks for the reviews @chunhtai and @justinmc. I was giving this a final look and may have noticed something I missed. 

In the test below when clicking between different selectables the selection is not cleared. Before this pr every click/tap would clear the selection, but now every click/tap is setting the selection. Should we be clearing the selection on the selectables that do not contain the click/tapped location? I'm leaning towards that we should, but what do you both think?

Some more context, on the native browser it seems like the selection is nulled for a node when clicking on another node, you can observe this behavior in this demo https://jsfiddle.net/mwt7fp6v/ . When clicking on a new node, in the case of the demo `<p>`, the other paragraphs do not contain a selection when using `Selection.containsNode()`, only the paragraph at the tapped location contains the collapsed selection.

```dart
    testWidgets('Test collapsing selection at multiple selectables', (WidgetTester tester) async {
      await tester.pumpWidget(
        MaterialApp(
          home: SelectableRegion(
            focusNode: FocusNode(),
            selectionControls: materialTextSelectionControls,
            child: const Column(
              children: <Widget>[
                Text('How are you?'),
                Text('Good, and you?'),
                Text('Fine, thank you.'),
              ],
            ),
          ),
        ),
      );
      final RenderParagraph paragraph1 = tester.renderObject<RenderParagraph>(find.descendant(of: find.text('How are you?'), matching: find.byType(RichText)));
      final TestGesture gesture = await tester.startGesture(textOffsetToPosition(paragraph1, 2), kind: PointerDeviceKind.mouse);
      addTearDown(gesture.removePointer);
      await tester.pump();
      await gesture.up();
      await tester.pumpAndSettle();
      expect(paragraph1.selections[0], const TextSelection.collapsed(offset: 2));

      final RenderParagraph paragraph2 = tester.renderObject<RenderParagraph>(find.descendant(of: find.text('Good, and you?'), matching: find.byType(RichText)));
      await gesture.down(textOffsetToPosition(paragraph2, 5));
      await tester.pump();
      await gesture.up();
      await tester.pumpAndSettle();
      expect(paragraph1.selections[0], const TextSelection.collapsed(offset: 12)); // should this be null?
      expect(paragraph2.selections[0], const TextSelection.collapsed(offset: 5));

      final RenderParagraph paragraph3 = tester.renderObject<RenderParagraph>(find.descendant(of: find.text('Fine, thank you.'), matching: find.byType(RichText)));
      await gesture.down(textOffsetToPosition(paragraph3, 13));
      await tester.pump();
      await gesture.up();
      await tester.pumpAndSettle();

      expect(paragraph1.selections[0], const TextSelection.collapsed(offset: 12)); // should this be null?
      expect(paragraph2.selections[0], const TextSelection.collapsed(offset: 14)); // should this be null?
      expect(paragraph3.selections[0], const TextSelection.collapsed(offset: 13));
    });
```

chunhtai commented:
It would be good if we can clear out selection in other unrelated selectable, but I think that may be hard to do? or may make the code messy.

If we don't clear our selection on unrelated selectable, there may be something we should think about.

1. This may affect how we are going to implement text cursor accessibility settings See https://support.google.com/chrome/answer/10129654?hl=en

Although I don't think this will be a problem because we have selectionGemeotry to tell us exactly where to paint the text cursor.

2. we may one day want to implement onSelectionChange in Text widget. Developer won't be able to tell whether the text does receive active collapsed selection vs no selection.


My suggestion would be if there is a clean way to clear out selection, then we should go for it.

Renzo-Olivares committed:
Add clearing unrelated selectable options

Renzo-Olivares commented:
@chunhtai Thank you for the suggestion! I agree that if we can clear the selection we probably should. I think there are a few ways we can approach this and have included some of them in this commit https://github.com/flutter/flutter/pull/132682/commits/8b33e9773fd4b3ce59ce245403fb13a426580229#diff-3c64e39c4f244a98f70ed8374eabfb2a4d4eeca07cd90a8f90abe4bc226bfe7f .

The options the following:
1. A new `CollapseSelectionSelectionEvent` (or `CollapseSelectionEvent`) . This will set both selection edges and clear the selection on all other selectables. Similar to `SelectWordSelectionEvent`.
2. Implement a private method to `clear all other unrelated selections /or selections that do not fall within the currentSelectionStartIndex and currentSelectionEndIndex range of the selection delegate`. It would be used like `_selectionDelegate.handleClearAllOtherSelections()` after we update the start/end edge to collapse it. This seems like a departure from dispatching selection events so this does not sound so clean. Another potential avenue we can take is create a new selection event `ClearAllOtherSelectionEvent` to do this but I am not sure if such an API makes sense.

Some options that do not work but I considered:
1. Calling `_clearSelection()` before updating the start/end edge to collapse the selection. This does clear the selection but after when we dispatch a `SelectionEdgeUpdateEvent` it dispatches to every selectable until reaching the target, so the selection is set once again in unrelated selectables.

Let me know what you think. Thanks!

chunhtai reviewed (commented):
If we would like to clear out other selectable, I think this would also apply to select word/line/paragraph.

> 1. A new CollapseSelectionSelectionEvent (or CollapseSelectionEvent) . This will set both selection edges and clear the selection on all other selectables. Similar to SelectWordSelectionEvent.

How do you clear other selectable? I imagine you will do something very similar to (2) to clear the selection. I think implementing 2 is fine as long as we keep the logic internal to MultiSelectableSelectionContainerDelegate and keep it self-contain. WDYT?

justinmc commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -962,6 +990,35 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
+  ///  * [_selectStartTo], which sets or updates selection start edge.
+  ///  * [_selectEndTo], which sets or updates selection end edge.
+  ///  * [_finalizeSelection], which stops the `continuous` updates.
+  ///  * [_clearSelection], which clear the ongoing selection.
```
"clear" => "clears"

justinmc commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -962,6 +990,35 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
+    // _selectEndTo(offset: offset);
+
+    // This works to clear all unrelated selections but relies on a new selection event.
+    _selectable?.dispatchSelectionEvent(CollapseSelectionSelectionEvent(globalPosition: offset));
```
Chun Heng should really have the final word, but I think dispatching a new event here seems fine...

justinmc reviewed (commented):
Overall the addition of the new selection event to clear the selections seems like a solid approach.

Renzo-Olivares commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -1715,11 +1717,30 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
   }
 
+  SelectionResult _handleCollapseSelection(Offset globalPosition) {
+    _selectableContainsOriginWord = true;
```
Good point! I'll keep this in mind. This block of code was pulled from `_handleSelectWord`s existing implementation so that probably needs a change.

Renzo-Olivares commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -1715,11 +1717,30 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
+    _selectableContainsOriginWord = true;
+
+    final TextPosition position = paragraph.getPositionForOffset(paragraph.globalToLocal(globalPosition));
+    if (_positionIsWithinCurrentSelection(position) && _textSelectionStart != _textSelectionEnd) {
```
Sorry about that, this block of code was pulled from `_handleSelectWord`s existing implementation so that case probably needs an adjustment or to be removed entirely.

Renzo-Olivares replied:
> This seems to be very similar to SelectWord, and probably SelectLine and SelectParagraph. What's the plan for expanding these functionality? An alternative is to use a single Selection event with different granularity

This will be addressed in the triple click gestures pr if we choose not to add a new selection event here.

Renzo-Olivares committed:
Revert "Add clearing unrelated selectable options"

This reverts commit 8b33e9773fd4b3ce59ce245403fb13a426580229.

Renzo-Olivares committed:
Add a private method to flush inactive selections

Renzo-Olivares committed:
rename

Renzo-Olivares committed:
Should flush inactive selections on setting selection edge

Renzo-Olivares commented:
> If we would like to clear out other selectable, I think this would also apply to select word/line/paragraph.
> 
> > 1. A new CollapseSelectionSelectionEvent (or CollapseSelectionEvent) . This will set both selection edges and clear the selection on all other selectables. Similar to SelectWordSelectionEvent.
> 
> How do you clear other selectable? I imagine you will do something very similar to (2) to clear the selection. I think implementing 2 is fine as long as we keep the logic internal to MultiSelectableSelectionContainerDelegate and keep it self-contain. WDYT?

Yes it would be implemented very similar to (2). I think that ideally setting the start and end edge to the same position would collapse it by default, and not leave any ongoing unrelated selections. What if we implement this private method, but instead of calling it in `_collapseSelectionAt` we call it in `_initSelection` and `_adjustSelection`. This way we will clear any inactive selections when setting the start/end edge. Something like this https://github.com/flutter/flutter/pull/132682/commits/936a2d5b97c55fded9975defbad49bcbf580b87b . If this does not seem reasonable then i'm okay calling the private method from `_collapseSelectionAt`. Let me know what you think.

chunhtai reviewed (commented):
> Yes it would be implemented very similar to (2). I think that ideally setting the start and end edge to the same position would collapse it by default, and not leave any ongoing unrelated selections. What if we implement this private method, but instead of calling it in _collapseSelectionAt we call it in _initSelection and _adjustSelection. This way we will clear any inactive selections when setting the start/end edge. Something like this https://github.com/flutter/flutter/commit/936a2d5b97c55fded9975defbad49bcbf580b87b . If this does not seem reasonable then i'm okay calling the private method from _collapseSelectionAt. Let me know what you think.

In which case would `_collapseSelectionAt` be invoked? Do you call it when receiving both selectStartAt and selectEndAt the same location back to back? Either way, this sounds fine to me.

Renzo-Olivares commented:
`_collapseSelectionAt` calls `selectStartTo` and `selectEndTo`. The `MultiSelectableContainerDelegate` is then responsible for flushing the unrelated selections when each edge is updated (when a SelectionEdgeUpdateEvent is dispatched).

chunhtai commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -2366,6 +2440,7 @@ abstract class MultiSelectableSelectionContainerDelegate extends SelectionContai
     } else {
       currentSelectionStartIndex = newIndex;
     }
+    _flushInactiveSelections();
```
just something to keep in mind, we know which selectables currently have selection after the adjustment, we need to go through the entire selectable list. This is fine as long as the performance is fine.

chunhtai reviewed (approved):
LGTM

Renzo-Olivares committed:
clear selection before collapsing it to ensure a previous selection is cleared

Renzo-Olivares committed:
update tests

Renzo-Olivares committed:
update tests

Renzo-Olivares committed:
updates to docs

Renzo-Olivares committed:
formatting fixes

Renzo-Olivares committed:
Make sure when we are adjusting an edge that the position we begin our walk through the selectable tree is inside the current viewport. If not we might walk through the tree incorrectly as the correct transformation calculations are not possible when the edge is not within the viewport

Renzo-Olivares committed:
update docs

Renzo-Olivares commented:
This PR is ready for another review. The changes since the last review are isolated to `MultiSelectableSelectionContainerDelegate._adjustSelection`. There was an issue where if we are adjusting a selection edge that is not inside the current viewport, the selectable tree walk would fail and the selection wouldn't be what the user expects. This is due to the transform matrix of a render object being set to the zero matrix when the render object is not visible, so any calculations done using this transform are invalid.

I have checked for this case, and attempt to start the tree walk at the selection edge that is visible in the current viewport / or start at 0 if neither edge is visible in the current viewport.

```dart
    final bool isCurrentEdgeWithinViewport = isEnd ? _selectionGeometry.endSelectionPoint != null : _selectionGeometry.startSelectionPoint != null;
    final bool isOppositeEdgeWithinViewport = isEnd ? _selectionGeometry.startSelectionPoint != null : _selectionGeometry.endSelectionPoint != null;
    // Determines if the edge being adjusted is within the current viewport.
    // If it is not, we attempt to locate the selection edge starting from the opposite end.
    // In cases where neither edge is within the viewport, the search for the new edge position
    // begins at 0.
    int newIndex = switch ((isEnd, isCurrentEdgeWithinViewport, isOppositeEdgeWithinViewport)) {
      (true, true, true) => currentSelectionEndIndex,
      (true, true, false) => currentSelectionEndIndex,
      (true, false, true) => currentSelectionStartIndex,
      (true, false, false) => 0,
      (false, true, true) => currentSelectionStartIndex,
      (false, true, false) => currentSelectionStartIndex,
      (false, false, true) => currentSelectionEndIndex,
      (false, false, false) => 0,
    };
```

chunhtai commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -2337,7 +2411,22 @@ abstract class MultiSelectableSelectionContainerDelegate extends SelectionContai
     }());
     SelectionResult? finalResult;
-    int newIndex = isEnd ? currentSelectionEndIndex : currentSelectionStartIndex;
+    final bool isCurrentEdgeWithinViewport = isEnd ? _selectionGeometry.endSelectionPoint != null : _selectionGeometry.startSelectionPoint != null;
```
Can you add some comment when this can be possible?

chunhtai reviewed (approved):
LGTM

Renzo-Olivares committed:
Address comments

Renzo-Olivares committed:
remember to dispose

Renzo-Olivares committed:
rerun Google Testing

Renzo-Olivares committed:
rerun Google Testing
